packages/boot/src/booters/model-api.booter.ts

```diff
@@ -0,0 +1,104 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/boot
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  config,
+  CoreBindings,
+  extensionPoint,
+  extensions,
+  Getter,
+  inject,
+} from '@loopback/core';
+import {
+  ModelApiBuilder,
+  ModelApiConfig,
+  MODEL_API_BUILDER_PLUGINS,
+} from '@loopback/model-api-builder';
+import {ApplicationWithRepositories, Model} from '@loopback/repository';
+import * as debugFactory from 'debug';
+import * as path from 'path';
+import {BootBindings} from '../keys';
+import {ArtifactOptions, booter} from '../types';
+import {BaseArtifactBooter} from './base-artifact.booter';
+
+const debug = debugFactory('loopback:boot:model-api');
+
+@booter('modelApi')
+@extensionPoint(MODEL_API_BUILDER_PLUGINS)
+export class ModelApiBooter extends BaseArtifactBooter {
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE)
+    public app: ApplicationWithRepositories,
+    @inject(BootBindings.PROJECT_ROOT) projectRoot: string,
+    @extensions()
+    public getModelApiBuilders: Getter<ModelApiBuilder[]>,
+    @config()
+    public booterConfig: ArtifactOptions = {},
+  ) {
+    // TODO assert that `app` has RepositoryMixin members
+
+    super(
+      projectRoot,
+      // Set booter options if passed in via bootConfig
+      Object.assign({}, RestDefaults, booterConfig),
+    );
+  }
+
+  async load(): Promise<void> {
```

Comment:
How do we make sure `model` classes are bound to the application context first during boot?

Reply:
> How do we make sure `model` classes are bound to the application context first during boot? Do we anticipate that model classes are bound during `configure` phase? Do we have to introduce a new `build` phase for booters?

Good catch!

Binding model classes during `configure` phase seems hacky to me. I was expecting that the model booter will be implemented the same way as other artifact booters, e.g. [RepositoryBooter](https://github.com/strongloop/loopback-next/blob/cc5cd20ae11e58c07a336774474ff6d7fe381a06/packages/boot/src/booters/repository.booter.ts).

Here are few options to consider:

(1)
Can we introduce a concept of dependencies to booters, allowing `ModelApiBooter` to request that `ModelBooter` is executed first?

```ts
@booter('modelApi', {dependsOn: [ModelBooter]})
@extensionPoint(MODEL_API_BUILDER_PLUGINS)
export class ModelApiBooter extends BaseArtifactBooter {
  // ...
}
```

(2)
Instead of introducing a Model booter, perhaps we can modify `ModelApiBooter` to load model classes, effectively merging Model + API booters into a single class.

- In the configuration, `ModelApiBooter` will require two sets of `ArtifactOptions` - one for model classes, another for configuration files.
- At discovery, the booter will discover both model and config files.
- When loading, we don't even have to bind model classes to context, API booter can look-up model classes from the array provided by `BaseArtifactBooter`.

(3)
Abandon the idea of booting model classes. Require the api-config object to provide a class constructor.

```ts
import {ModelApiConfig} from '@loopback/rest-crud';
import {Todo} from '../models';

module.exports = <ModelApiConfig>{
  model: Todo,
  pattern: 'CrudRest',
  dataSource: 'db',
  basePath: '/todos',
};
```

This may simplify consumption of models from 3rd-party packages (npm modules) - instead of having to tweak booter configuration, developers can decide in each config file where to load the model from.

```ts
import {Todo} from '@acme/process-models';
import {ModelApiConfig} from '@loopback/rest-crud';

module.exports = <ModelApiConfig>{
  model: Todo,
  pattern: 'CrudRest',
  dataSource: 'db',
  basePath: '/todos',
};
```

---

Personally, I am leaning towards the last option, as it's the easier to implement. We can look into other options later, after the v1 implementation has been shipped.

Thoughts?

/cc @strongloop/loopback-next @strongloop/loopback-maintainers

Reply:
I have updated the spike with the 3rd option, The ModelApiConfig object is referencing the model class directly now, setting model field to an imported class constructor.